x/vote/manager/manager.go

```diff
@@ -0,0 +1,294 @@
+	// withdraw money from voter's bank
+	if err := vm.am.MinusCoinFromUsername(ctx, username, amount); err != nil {
+		return err
+	}
+	return vm.AddStake(ctx, username, amount)
+}
+
+func (vm VoteManager) AddStake(ctx sdk.Context, username linotypes.AccountKey, amount linotypes.Coin) sdk.Error {
+	voter, err := vm.storage.GetVoter(ctx, username)
+	if err != nil {
+		if err.Code() != linotypes.CodeVoterNotFound {
+			return err
+		}
+		voter = &model.Voter{
+			Username:          username,
+			LinoStake:         linotypes.NewCoinFromInt64(0),
+			LastPowerChangeAt: ctx.BlockHeader().Time.Unix(),
+			Duty:              types.DutyVoter,
+			Interest:          linotypes.NewCoinFromInt64(0),
+		}
+	}
+
+	interest, err := vm.gm.GetInterestSince(ctx, voter.LastPowerChangeAt, voter.LinoStake)
+	if err != nil {
+		return err
+	}
+	voter.Interest = voter.Interest.Plus(interest)
+	voter.LinoStake = voter.LinoStake.Plus(amount)
+	voter.LastPowerChangeAt = ctx.BlockHeader().Time.Unix()
+
+	if err := vm.storage.SetVoter(ctx, username, voter); err != nil {
+		return err
+	}
+	// add linoStake to global stat
+	if err := vm.gm.AddLinoStakeToStat(ctx, amount); err != nil {
+		return err
+	}
+	return vm.AfterAddingStake(ctx, username)
+}
+
+func (vm VoteManager) StakeOut(ctx sdk.Context, username linotypes.AccountKey, amount linotypes.Coin) sdk.Error {
+	if err := vm.MinusStake(ctx, username, amount); err != nil {
+		return err
+	}
+
+	param, err := vm.paramHolder.GetVoteParam(ctx)
+	if err != nil {
+		return err
+	}
+
+	if err := vm.am.AddFrozenMoney(
+		ctx, username, amount, ctx.BlockHeader().Time.Unix(),
+		param.VoterCoinReturnIntervalSec, param.VoterCoinReturnTimes); err != nil {
+		return err
+	}
+
+	events, err := accmn.CreateCoinReturnEvents(
+		ctx, username, param.VoterCoinReturnTimes, param.VoterCoinReturnIntervalSec, amount, linotypes.VoteReturnCoin)
+	if err != nil {
+		return err
+	}
+
+	if err := vm.gm.RegisterCoinReturnEvent(
+		ctx, events, param.VoterCoinReturnTimes, param.VoterCoinReturnIntervalSec); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (vm VoteManager) MinusStake(ctx sdk.Context, username linotypes.AccountKey, amount linotypes.Coin) sdk.Error {
+	voter, err := vm.storage.GetVoter(ctx, username)
+	if err != nil {
+		return err
+	}
+
+	// make sure stake is sufficient excludes frozen amount
+	if !voter.LinoStake.Minus(voter.FrozenAmount).IsGTE(amount) {
+		return types.ErrInsufficientStake()
+	}
+
+	interest, err := vm.gm.GetInterestSince(ctx, voter.LastPowerChangeAt, voter.LinoStake)
+	if err != nil {
+		return err
+	}
+	voter.Interest = voter.Interest.Plus(interest)
+	voter.LinoStake = voter.LinoStake.Minus(amount)
+	voter.LastPowerChangeAt = ctx.BlockHeader().Time.Unix()
+
+	if err := vm.storage.SetVoter(ctx, username, voter); err != nil {
+		return err
+	}
+	// add linoStake to global stat
+	if err := vm.gm.MinusLinoStakeFromStat(ctx, amount); err != nil {
+		return err
+	}
+	return vm.AfterSubtractingStake(ctx, username)
```

Comment:
So it is possible that stake actually increase when call AfterSubtractingStake since add interests?